### ISSUE TITLE
False positive in `array_init`

### ISSUE THREAD

**Issue by liam-i** [realm/SwiftLint#3749]
### New Issue Checklist

- [X] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint lint
/Users/liam/Documents/Example/AppDelegate.swift:26:45: Array Init Violation: Prefer using `Array(seq)` over `seq.map { $0 }` to convert a sequence into an Array. (array_init)
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)? 0.45.0
* Installation method used (Homebrew, CocoaPods, building from source, etc)? CocoaPods
* Paste your configuration file:

```yml
only_rules:
  - array_init
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)?
  If so, paste their relative paths and respective contents.
No

* Which Xcode version are you using (check `xcodebuild -version`)?
Xcode 13.1

* Do you have a sample that shows the issue? Run `echo "[string here]" | swiftlint lint --no-cache --use-stdin --enable-all-rules`
  to quickly test if your example is really demonstrating the issue. If your example is more
  complex, you can use `swiftlint lint --path [file here] --no-cache --enable-all-rules`.

```swift
enum BizError: Swift.Error {
}

let reqResult: Result<String, BizError> = .success("")

// False positive in `array_init`
let result: Result<Any, BizError> = reqResult.map { $0 }

print(result)
```

**Comment by aranasaurus** [comment]
I spent some time looking at this today, and I could be wrong since this is my first time jumping into the codebase, but it looks like the way to fix this issue would require changing the ArrayInitRule's `KindType` to `SwiftDeclarationKind`, but that would make this rule stop being able to detect occurrences of `.map` instead of `Array()` as arguments to function calls.

I couldn't see a way to make one ASTRule able to operate on more than one KindType. Maybe there's another way that I'm just missing?

**Comment by SimplyDanny** [comment]
My take on this is that it cannot be fixed without knowing the type of the object at which `.map` is called. By default, SwiftLint deals with an AST that is not type checked. I only recently learned that there are a few so called "Analyzer Rules" which have a type checked AST available to work on. They are currently an experimental feature, though.

**Comment by jpsim** [comment]
> They are currently an experimental feature, though.

Analyzer rules have been pretty stable and we could remove the "experimental" label on them.

**Comment by jpsim** [comment]
https://github.com/realm/SwiftLint/pull/3884

**Comment by SimplyDanny** [comment]
So, is it okay to convert the `array_init` rule to an analysis rule to fix the bug, @jpsim? If yes, I could have a look into this.

**Comment by SimplyDanny** [comment]
That's what I also thought after reading your comment. Let me give it a try.

**Comment by SimplyDanny** [comment]
Can this one be closed now that there is TypesafeArrayInitRule available, @jpsim? I mean, the bug is still valid but certainly won't be fixed.

**Comment by jpsim** [comment]
Yes let's close this, if someone wants a type-checked version of the rule they can use TypesafeArrayInitRule, albeit requiring the compilation database.
